### ISSUE TITLE
[Bug?]: illegal operation on a directory when using package.json scripts to a shell script

### ISSUE THREAD

**Issue by a666** [yarnpkg/berry#5669]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

In version 3.6.2 having a packaje.json script that runs a shell script sends the following in version 3.6.2:
```shell
$ yarn compose
Internal Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:751:3)
    at $t.readSync (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:314:1260)
    at zR (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:423:663)
    at LB (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:423:1888)
    at async m9 (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:423:1934)
    at async wRe (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:422:3297)
    at async /home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:422:2583
    at async $t.mktempPromise (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:314:69436)
    at async Object.TB (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:422:2523)
    at async Fu.execute (/home/***/.cache/node/corepack/yarn/3.6.2/yarn.js:572:2017)
```
but in version 3.6.1 the script just runs
```shell
$ yarn compose
# prints the help message
Runs a ...
Usage: yarn run compose [-h] [-f]
```

### To reproduce

Having the following `package.json`:
```json
{
  "name": "test",
  "private": true,
  "packageManager": "yarn@3.6.2",
  "scripts": {
    "compose": "./compose.sh",
  }
}
```
And the following `compose.sh`
```bash
#!/usr/bin/env bash

echo Do you feel lucky punk?
```

Should be enough.

PS. BTW the link to https://yarnpkg.com/advanced/sherlock is broken.

### Environment

```shell
System:
    OS: Linux 6.4 Arch Linux
    CPU: (12) x64 Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
  Binaries:
    Node: 18.17.1 - /tmp/xfs-7f8286f6/node
    Yarn: 3.6.2 - /tmp/xfs-7f8286f6/yarn
    npm: 9.6.7 - ~/.cache/node/bin/npm
```

### Additional context

Considering how simple  this pattern is, probably will happen to any package.json script that runs any executable script (like php's vendor/bin/some_package_or_another).

**Comment by la-magra** [comment]
Isn't this a breaking change?

**Comment by danielpza** [comment]
I'm also seeing this issue on yarn rc v4 : `yarn-4.0.0-rc.50`

```
Internal Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:749:3)
    at Rn.readSync ([...]/.yarn/releases/yarn-4.0.0-rc.50.cjs:9:7517)
...
```

**Comment by DBosley** [comment]
I'm running into this issue with the latest versions of yarn (3.6.2 and 3.6.3) on a workspace enabled monorepo. The workspaces has the following dependency:
`"mini-css-extract-plugin": "0.4.5"`
I've traced the error in the yarn-3.6.3.cjs file. It seems to think mini-css-extract-pugin is a binary, but it's a directory.

I put a console log in this block of minified code (you can search for it in the yarn-3.6.3.cjs file)
`{let C=x.resolve(f,p);i.set(h,[g,K.fromPortablePath(C),zR(C)])}`
I changed it to:
`{let C=x.resolve(f,p);console.log(C);`i.set(h,[g,K.fromPortablePath(C),zR(C)])}`

This printed out `node_modules/mini-css-extract-plugin` in the console after a bunch of runnable files from the .bin folder. The rest of the outputs from this `C` variable were all files, while this was a folder.


My current workaround is to use 3.6.1 which does not seem to have the issue.

**Comment by DBosley** [comment]
Oh, I did also notice this in the yarn.lock
```
"mini-css-extract-plugin@npm:^0.4.1":
  version: 0.4.5
  resolution: "mini-css-extract-plugin@npm:0.4.5"
  dependencies:
    loader-utils: ^1.1.0
    schema-utils: ^1.0.0
    webpack-sources: ^1.1.0
  peerDependencies:
    webpack: ^4.4.0
  bin:
    mini-css-extract-plugin: ""
  checksum: 0e8722b879e078cbc4d31a6c3aaf410ccc185d12e4f414301c0c4586a8e0c5d56e39bd281a268e0324cc9c79cfe8a29b27d4cc02afaf805926d8c1769f9e40fe
  languageName: node
  linkType: hard
  ```
  notice the line saying it's a bin... That's definitely not right.

**Comment by rally25rs** [comment]
Was finally able to reproduce this;

* node v18.17.1
* corepack enabled

`package.json`
```json
{
  "name": "test",
  "packageManager": "yarn@3.6.2",
  "scripts": {
    "test": "echo 'test'"
  },
  "dependencies": {
    "mini-css-extract-plugin": "0.4.5"
  }
}
```

`.yarnrc.yml`
```yaml
nodeLinker: 'node-modules'
```

```
$ yarn run test

Internal Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:751:3)
    at $t.readSync (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:314:1260)
    at zR (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:423:663)
    at LB (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:423:1888)
    at async m9 (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:423:1934)
    at async wRe (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:422:3297)
    at async /Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:422:2583
    at async $t.mktempPromise (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:314:69436)
    at async Object.TB (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:422:2523)
    at async Fu.execute (/Users/jeff/.cache/node/corepack/yarn/3.6.2/yarn.js:572:2017)
```

This seems to only happen for me if using the `n_m` linker; `pnp` does not throw an error.

---

What seems to cause the issue to show up is having an incorrectly formatted dependency that has an invalid `bin` entry in its `package.json`

For example `mini-css-extract-plugin@0.4.5` has `"bin": ""` which causes yarn to append this to the `.bin` path to make a path to the executable script, but appending empty string just results in the directory name itself, hence the "illegal operation on a directory" error.

---

As a fix, It seems like yarn could do some basic validation on the `bin` entry and skip linking it if it looks invalid.

**Comment by a666** [comment]
I would kindly request reopening this issue.

#5743 didn't address the original problem (which relates to using package.json scripts to call to any kind of shell/php/python scripts).

**Comment by Vages** [comment]
I have opened a similar issue: https://github.com/yarnpkg/berry/issues/5766 (Thanks for your comments, @DBosley and @rally25rs; they really helped me figure out what _my_ issue was.)

Your minimal example works in both Yarn 3.6.4 and 3.6.2 on my machine, @a666. At first it didn’t, because there’s a syntax error in the JSON file you posted (I’m surprised how you got it to work):

```shell
yarn run compose
Syntax Error: Unexpected token } in JSON at position 121 (when parsing /path/to/dir/package.json)
```

There should be no comma at the end of the `compose`, as it is the last item in its object. Here is the diff:

```diff
6c6
<     "compose": "./compose.sh",
---
>     "compose": "./compose.sh"
```

After that, I could run the following commands:
 
```shell
$ yarn set version 3.6.2 # Install Yarn version to .yarn folder
$ yarn install
$ yarn run compose
Do you feel lucky punk?
```

I even repeated the process above, replacing 3.6.2 with 3.6.4, and that worked out all right.

Here are some suggestions as to what may be the problem. you can try:

- *The `package.json` you have posted has been edited compared to the last time `yarn.lock` was updated.* Maybe there is some dependency left in there that triggers the same bug I found.
- *`corepack` is inadvertently parsing some other `yarn.lock` file than the one you have in your project,* possibly triggering the same bug I found. 
    - Try running `corepack disable` and run the commands that I ran to get things to work.

**Comment by a666** [comment]
Yeah, it took us a while to track this to `jsbarcode` too.
(we have ~600 deps, massive monorepo, no comments of any kind in package management files, pure hell).
It seems unrelated to related to corepack.

This issue can remain closed in favor of #5766.

Sorry for the deficient report.
